internal/memcmd/observer.go

```diff
@@ -0,0 +1,87 @@
+package memcmd
+
+import (
+	"sync"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// Observer is an interface for observing and tracking the memory usage of a process.
+//
+// Implementations of this interface should provide methods to start and stop the observation,
+// as well as retrieve the maximum memory usage of the observed process.
+//
+// Callers must call Stop when they are done with the observer to release any associated resources.
+type Observer interface {
+	// Start starts the observer. It should be called before any other method.
+	//
+	// After Start is called, callers must call Stop when they are done with the
+	// observer to release any resources.
+	//
+	// Calling Start() multiple times is safe and has no effect after the first invocation.
+	Start()
+
+	// Stop stops the observer and releases any associated resources.
```

Comment:
Maybe to clarify that is how it's meant to be used? 

```suggestion
	// Stop stops the observer and releases any associated resources. For accurate measurement,
	// Stop must be called _after_ Wait has been called on the *exec.Cmd.
```